### DISCUSSION TITLE
Cannot inject custom http client for text to audio and audio to text services using open ai services

### DISCUSSION THREAD

**Discussion by giannik** [microsoft/semantic-kernel#9277] [Q&A] [answered]
Semantic kernel allows for open AI chat completion to provide your own HTTP client in the constructor while building the kernel . This is very useful for multi tenant scenarios to better organize http client management.
But I noticed that for open ai audio to text and open AI text to audio the  constructor  does not allow this implementation for custom HTTP client. Why is this?

```
  //this works 
Kernel.CreateBuilder()
    .AddOpenAIChatCompletion("gpt-4-vision-preview", decryptedOpenAiKey , httpclient )   //this works 




//but for these services i cannot inject own http client (only openaiclient is allowed.) . i wish to follow same pattern as chat   completion
builder.Services.AddOpenAIAudioToText(
    modelId: AudioToTextModel,
    apiKey: decryptedOpenAiKey, ???
    );

builder.Services.AddOpenAITextToAudio(
    modelId: TextToAudioModel,
    apiKey: decryptedOpenAiKey, ???
   );
```

**Comment by evchaki** [comment]
@giannik - thanks for asking.  @RogerBarret0 - can you provide some guidance on why we did this?

**Comment by RogerBarret0** [comment, marked as answer]
When using `builder.Services` you are targeting `ServiceCollection` which is a Dependency Injection focused collection.

To achieve your goal, you need to have a `HttpClient` added to your services DI Container so it can be injected automatically in the Connector Services.

`AddOpenAITextToAudio` code.
https://github.com/microsoft/semantic-kernel/blob/e23d6364578dcd40ce380b57e176a48d3c7c3e61/dotnet/src/Connectors/Connectors.OpenAI/Extensions/OpenAIServiceCollectionExtensions.cs#L179

`HttpClientProvider.GetHttpClient`
https://github.com/microsoft/semantic-kernel/blob/e23d6364578dcd40ce380b57e176a48d3c7c3e61/dotnet/src/InternalUtilities/src/Http/HttpClientProvider.cs#L39

**Comment by giannik** [comment]
@RogerBarret0  thank you. 
Does'nt this mean that from the service collection Semantic kernel will pick the first available HttpClient available  ?

What if i have injected more httpclients for other purposes or if the service collection comes from my main app and not specific to semantic kernel. 

Maybe if somehow i could specify  with a key that this http client is specific for semantic kernel only ?
